### PR TITLE
Issue5423 and issue5465

### PR DESCRIPTION
**Pull request by neil4dong**
BugFix：
1.https://github.com/apache/shardingsphere/issues/5210  
ON DUPLICATE KEY UPDATE Sub-Clause paramter dropped   

BugFix: 
2.https://github.com/apache/shardingsphere/issues/5465
1)ON DUPLICATE KEY UPDATE Sub-Clause encrypt logic: missing assistedQueryColumn.
2)REMOVE wrong implements of QueryWithCipherColumnAware in  org.apache.shardingsphere.encrypt.rewrite.parameter.impl.EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter

### CODE REVIEW

**Comment by neil4dong** [comment]
> Hi, can you split them to 2 pull requests?
> Review a big pull request is too difficult.

@tristaZero and I discussed this before ,  These two issues are strongly connected.  I cannot fix one without changing another' code.   If I force to do so , I can only change the issue #5465 from "not work properly " to "another way of not work properly".   :)

**Comment by tristaZero** [comment]
@terrymanu  I know your concern. However, there is a close relationship between two Issues. I will go through it carefully.

File: shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/GenericSQLRewriteEngine.java [hunk 1]
```diff
@@ -34,6 +40,22 @@
...
      */
     public GenericSQLRewriteResult rewrite(final SQLRewriteContext sqlRewriteContext) {
-        return new GenericSQLRewriteResult(new SQLRewriteUnit(new DefaultSQLBuilder(sqlRewriteContext).toSQL(), sqlRewriteContext.getParameterBuilder().getParameters()));
+        return new GenericSQLRewriteResult(new SQLRewriteUnit(new DefaultSQLBuilder(sqlRewriteContext).toSQL(), getParameters(sqlRewriteContext.getParameterBuilder())));
+    }
+    
+    private List<Object> getParameters(final ParameterBuilder parameterBuilder) {
+        if (parameterBuilder instanceof StandardParameterBuilder) {
+            return parameterBuilder.getParameters();
+        } else {
```
Comment (tristaZero) [inline shardingsphere-underlying/shardingsphere-rewrite/shardingsphere-rewrite-engine/src/main/java/org/apache/shardingsphere/underlying/rewrite/engine/GenericSQLRewriteEngine.java:49]: It is preferable to move the content in `else {}` to the parent level.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/statement/impl/InsertStatementContextTest.java [hunk 2]
```diff
@@ -79,6 +98,22 @@ private void setUpInsertValues(final InsertStatement insertStatement) {
...
+    private void setUpOnDuplicateValues(final InsertStatement insertStatement) {
+        AssignmentSegment parameterMarkerExpressionAssignment = new AssignmentSegment(0, 0,
+                new ColumnSegment(0, 0, new IdentifierValue("on_duplicate_key_update_column_1")),
+                new ParameterMarkerExpressionSegment(0, 0, 4)
+        );
+        AssignmentSegment literalExpressionAssignment = new AssignmentSegment(0, 0,
+                new ColumnSegment(0, 0, new IdentifierValue("on_duplicate_key_update_column_2")),
+                new LiteralExpressionSegment(0, 0, 5)
+        );
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/statement/impl/InsertStatementContextTest.java:110]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        Collection<AssignmentSegment> assignments = Lists.newArrayList();
+        List<Object> parameters = Collections.emptyList();
+        int parametersOffset = 0;
+
+        OnDuplicateUpdateContext onDuplicateUpdateContext = new OnDuplicateUpdateContext(assignments, parameters, parametersOffset);
+
+        Method getParameterIndexMethod = OnDuplicateUpdateContext.class.getDeclaredMethod("getParameterIndex", ExpressionSegment.class);
+        getParameterIndexMethod.setAccessible(true);
+        ParameterMarkerExpressionSegment notExistsExpressionSegment = new ParameterMarkerExpressionSegment(0, 0, 0);
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:127]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    
+    @Test
+    public void assertGetParameterIndex() throws NoSuchMethodException, IllegalAccessException {
+        Collection<AssignmentSegment> assignments = Lists.newArrayList();
+        List<Object> parameters = Collections.emptyList();
+        int parametersOffset = 0;
+
+        OnDuplicateUpdateContext onDuplicateUpdateContext = new OnDuplicateUpdateContext(assignments, parameters, parametersOffset);
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:123]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        ColumnSegment column = new ColumnSegment(doesNotMatterLexicalIndex, doesNotMatterLexicalIndex, new IdentifierValue(doesNotMatterColumnName));
+        return new AssignmentSegment(doesNotMatterLexicalIndex, doesNotMatterLexicalIndex, column, expressionSegment);
+    }
+    
+    @Test
+    public void assertGetParameterIndex() throws NoSuchMethodException, IllegalAccessException {
+        Collection<AssignmentSegment> assignments = Lists.newArrayList();
+        List<Object> parameters = Collections.emptyList();
+        int parametersOffset = 0;
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:121]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    private Collection<AssignmentSegment> makeLiteralExpressionSegment(final Object literalObject) {
+        LiteralExpressionSegment parameterLiteralExpression = new LiteralExpressionSegment(0, 10, literalObject);
+        AssignmentSegment assignmentSegment = makeAssignmentSegment(parameterLiteralExpression);
+        return Collections.singleton(assignmentSegment);
+    }
+    
+    private AssignmentSegment makeAssignmentSegment(final SimpleExpressionSegment expressionSegment) {
+        int doesNotMatterLexicalIndex = 0;
+        String doesNotMatterColumnName = "columnNameStr";
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:111]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        assertThat(valueFromInsertValueContext1, is(parameterValue1));
+
+        Object valueFromInsertValueContext2 = onDuplicateUpdateContext.getValue(1);
+        assertThat(valueFromInsertValueContext2, is(parameterValue2));
+    }
+    
+    private Collection<AssignmentSegment> makeParameterMarkerExpressionAssignmentSegment() {
+        ParameterMarkerExpressionSegment parameterMarkerExpressionSegment = new ParameterMarkerExpressionSegment(0, 10, 5);
+        AssignmentSegment assignmentSegment1 = makeAssignmentSegment(parameterMarkerExpressionSegment);
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:86]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        Method calculateParametersCountMethod = OnDuplicateUpdateContext.class.getDeclaredMethod("calculateParametersCount", Collection.class);
+        calculateParametersCountMethod.setAccessible(true);
+        int calculateParametersCountResult = (int) calculateParametersCountMethod.invoke(onDuplicateUpdateContext, new Object[]{assignments});
+        assertThat(onDuplicateUpdateContext.getParametersCount(), is(calculateParametersCountResult));
+
+        Method getValueExpressionsMethod = OnDuplicateUpdateContext.class.getDeclaredMethod("getValueExpressions", Collection.class);
+        getValueExpressionsMethod.setAccessible(true);
+        List<ExpressionSegment> getValueExpressionsResult = (List<ExpressionSegment>) getValueExpressionsMethod.invoke(onDuplicateUpdateContext, new Object[]{assignments});
+        assertThat(onDuplicateUpdateContext.getValueExpressions(), is(getValueExpressionsResult));
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:61]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        List<Object> parameters = Collections.emptyList();
+        int parametersOffset = 0;
+
+        OnDuplicateUpdateContext onDuplicateUpdateContext = new OnDuplicateUpdateContext(assignments, parameters, parametersOffset);
+
+        Method calculateParametersCountMethod = OnDuplicateUpdateContext.class.getDeclaredMethod("calculateParametersCount", Collection.class);
+        calculateParametersCountMethod.setAccessible(true);
+        int calculateParametersCountResult = (int) calculateParametersCountMethod.invoke(onDuplicateUpdateContext, new Object[]{assignments});
+        assertThat(onDuplicateUpdateContext.getParametersCount(), is(calculateParametersCountResult));
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:56]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java [hunk 10]
```diff
@@ -93,9 +114,21 @@ public InsertStatementContext(final SchemaMetaData schemaMetaData, final List<Ob
...
     
+    /**
+     * Get OnDuplicateKeyUpdateParameters.
+     * @return OnDuplicateKeyUpdateParameters
+     */
+    public List<Object> getOnDuplicateKeyUpdateParameters() {
+        if (null == onDuplicateKeyUpdateValueContext) {
+            return new ArrayList<>(0);
+        }
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java:125]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java [hunk 11]
```diff
@@ -50,30 +54,47 @@
...
+    private Optional<OnDuplicateUpdateContext> getOnDuplicateKeyUpdateValueContext(final List<Object> parameters, final AtomicInteger parametersOffset) {
+
+        if (!getSqlStatement().getOnDuplicateKeyColumns().isPresent()) {
+            return Optional.empty();
+        }
+        Collection<AssignmentSegment> onDuplicateKeyColumns = getSqlStatement().getOnDuplicateKeyColumns().get().getColumns();
+        OnDuplicateUpdateContext onDuplicateUpdateContext = new OnDuplicateUpdateContext(onDuplicateKeyColumns, parameters, parametersOffset.get());
+        parametersOffset.addAndGet(onDuplicateUpdateContext.getParametersCount());
+        return Optional.of(onDuplicateUpdateContext);
+
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java:92]: Please remove this redundant blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java [hunk 12]
```diff
@@ -50,30 +54,47 @@
...
         super(sqlStatement);
         tablesContext = new TablesContext(sqlStatement.getTable());
         columnNames = sqlStatement.useDefaultColumns() ? schemaMetaData.getAllColumnNames(sqlStatement.getTable().getTableName().getIdentifier().getValue()) : sqlStatement.getColumnNames();
-        insertValueContexts = getInsertValueContexts(parameters);
+
+        AtomicInteger parametersOffset = new AtomicInteger(0);
+        insertValueContexts = getInsertValueContexts(parameters, parametersOffset);
+        onDuplicateKeyUpdateValueContext = getOnDuplicateKeyUpdateValueContext(parameters, parametersOffset).orElse(null);
+
         generatedKeyContext = new GeneratedKeyContextEngine(schemaMetaData).createGenerateKeyContext(parameters, sqlStatement).orElse(null);
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java:70]: Please remove this redundant blank line.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java [hunk 13]
```diff
@@ -316,6 +320,79 @@ public void assertAddBatchWithGenerateKeyColumn() throws SQLException {
...
+    @Test
+    public void assertAddOnDuplicateKey() throws SQLException {
+        int itemId = 1;
+        int userId1 = 101;
+        int userId2 = 102;
+        int orderId = 200;
+        String status = "init";
+        
+        String updatedStatus = "updated on duplicate key";
+
```
Comment (tristaZero) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:332]: Please remove this redundant blank line.

File: encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java [hunk 14]
```diff
@@ -51,30 +47,36 @@ protected boolean isNeedRewriteForEncrypt(final SQLStatementContext sqlStatement
...
+                        groupedParameterBuilder.getOnDuplicateKeyUpdateParametersBuilder().getAddedIndexAndParameters().put(columnIndex + 1, new LinkedList<>());
+                    }
+                    groupedParameterBuilder.getOnDuplicateKeyUpdateParametersBuilder().getAddedIndexAndParameters().get(columnIndex + 1).addAll(addedParameters);
                 }
-            }
-            if (null != plainColumnValue) {
-                groupedParameterBuilder.getOnDuplicateKeyUpdateAddedParameters().add(plainColumnValue);
-            }
+            });
+
```
Comment (tristaZero) [inline encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java:79]: Please remove this redundant blank line.

File: encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java [hunk 15]
```diff
@@ -51,30 +47,36 @@ protected boolean isNeedRewriteForEncrypt(final SQLStatementContext sqlStatement
     @Override
     public void rewrite(final ParameterBuilder parameterBuilder, final InsertStatementContext insertStatementContext, final List<Object> parameters) {
         String tableName = insertStatementContext.getSqlStatement().getTable().getTableName().getIdentifier().getValue();
-        Preconditions.checkState(insertStatementContext.getSqlStatement().getOnDuplicateKeyColumns().isPresent());
-        OnDuplicateKeyColumnsSegment onDuplicateKeyColumnsSegment = insertStatementContext.getSqlStatement().getOnDuplicateKeyColumns().get();
-        Collection<AssignmentSegment> onDuplicateKeyColumnsSegments = onDuplicateKeyColumnsSegment.getColumns();
-        if (onDuplicateKeyColumnsSegments.isEmpty()) {
-            return;
-        }
+
```
Comment (tristaZero) [inline encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java:50]: Please remove this redundant blank line.

**Comment by tristaZero** [comment]
Very appreciate your effort! 👍 
It made this PR notable that clear coding logic and sufficient test cases. However, in order to make it top-notch, i left my comments here, which mainly involve the check style. So do not worry, it is easy to handle, IMO. 😉 
Look forward to your response, @neil4dong

**Comment by neil4dong** [comment]
> Very appreciate your effort! 👍
> It made this PR notable that clear coding logic and sufficient test cases. However, in order to make it top-notch, i left my comments here, which mainly involve the check style. So do not worry, it is easy to handle, IMO. 😉
> Look forward to your response, @neil4dong

All resolved to meet the code style we used in this project 😉

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java (hunk 9 above)
Comment (neil4dong) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:55]: IMO,it's better to keep the different method call with assertion separately.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java (hunk 8 above)
Comment (neil4dong) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/test/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContextTest.java:60]: IMO,it's better to keep the different method call with assertion separately.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java [hunk 16]
```diff
@@ -126,7 +130,7 @@ public void assertMultiValuesWithGenerateShardingKeyColumn() throws SQLException
             }
         }
     }
-
+
```
Comment (terrymanu) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:133]: Please keep original indent

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java [hunk 17]
```diff
@@ -83,7 +87,7 @@ public void assertAddBatch() throws SQLException {
...
```
Comment (terrymanu) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:90]: Please keep original indent

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java [hunk 18]
```diff
@@ -316,6 +320,77 @@ public void assertAddBatchWithGenerateKeyColumn() throws SQLException {
...
+            preparedStatement.setInt(3, userId1);
+            preparedStatement.setString(4, status);
+            preparedStatement.setInt(5, itemId);
+            preparedStatement.setInt(6, orderId);
+            preparedStatement.setInt(7, userId2);
+            preparedStatement.setString(8, status);
+            preparedStatement.setString(9, updatedStatus);
+            int result = preparedStatement.executeUpdate();
+            assertThat(result, is(2));
+
```
Comment (terrymanu) [inline sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:345]: Please remove useless blank lines.
If we think use blank line to separate diff logic block, it is better extract them as a new private method.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java [hunk 19]
```diff
@@ -93,9 +110,20 @@ public InsertStatementContext(final SchemaMetaData schemaMetaData, final List<Ob
         return result;
     }
     
+    /**
+     * Get OnDuplicateKeyUpdateParameters.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/statement/dml/InsertStatementContext.java:114]: Please fix javadoc, use words or {@code } to label class name

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java (hunk 16 above)
Comment (neil4dong) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:133]: It's wired , there is two kinds of ident, some files no indent ,some files has ident .  which one should I use?

Comment (terrymanu) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:133]: we prefer just keep 4 spaces

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContext.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            if (each instanceof ParameterMarkerExpressionSegment) {
+                result++;
+            }
+        }
+        throw new IllegalArgumentException("Can not get parameter index.");
+    }
+    
+    /**
+     * get on duplicate key update column by index of this clause.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContext.java:103]: Please use upper case for the first letter of java doc.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContext.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+     * @param index index.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContext.java:104]: please remove `.` in `@param`

Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-binder/src/main/java/org/apache/shardingsphere/sql/parser/binder/segment/insert/values/OnDuplicateUpdateContext.java:104]: please keep one line between with java doc and `@param`

**Comment by neil4dong** [comment]
All clear.     Please imporve checkstyle. It is really time consuming.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java (hunk 17 above)
Comment (neil4dong) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:90]: use 4 spaces.

File: sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java (hunk 16 above)
Comment (neil4dong) [reply sharding-jdbc/sharding-jdbc-core/src/test/java/org/apache/shardingsphere/shardingjdbc/jdbc/core/statement/ShardingSpherePreparedStatementTest.java:133]: use 4 spaces.

**Comment by tristaZero** [comment]
@neil4dong Hi Congratulation! I saw your ID name in contributors list, terrific😉

**Pull request merged**